shell/renderer/api/atom_api_context_bridge.cc

```diff
@@ -0,0 +1,473 @@
+// Copyright (c) 2019 Slack Technologies, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "shell/renderer/api/atom_api_context_bridge.h"
+
+#include <set>
+#include <utility>
+#include <vector>
+
+#include "base/strings/string_number_conversions.h"
+#include "shell/common/api/remote/object_life_monitor.h"
+#include "shell/common/native_mate_converters/blink_converter.h"
+#include "shell/common/native_mate_converters/callback_converter_deprecated.h"
+#include "shell/common/native_mate_converters/once_callback.h"
+#include "shell/common/promise_util.h"
+
+namespace electron {
+
+namespace api {
+
+namespace {
+
+static int kMaxRecursion = 10000;
+
+content::RenderFrame* GetRenderFrame(const v8::Local<v8::Object>& value) {
+  v8::Local<v8::Context> context = value->CreationContext();
+  if (context.IsEmpty())
+    return nullptr;
+  blink::WebLocalFrame* frame = blink::WebLocalFrame::FrameForContext(context);
+  if (!frame)
+    return nullptr;
+  return content::RenderFrame::FromWebFrame(frame);
+}
+
+std::map<content::RenderFrame*, context_bridge::RenderFramePersistenceStore*>
+    store_map_;
+
+context_bridge::RenderFramePersistenceStore* GetOrCreateStore(
+    content::RenderFrame* render_frame) {
+  auto it = store_map_.find(render_frame);
+  if (it == store_map_.end()) {
+    auto* store = new context_bridge::RenderFramePersistenceStore(render_frame);
+    store_map_[render_frame] = store;
+    return store;
+  }
+  return it->second;
+}
+
+// Sourced from "extensions/renderer/v8_schema_registry.cc"
+// Recursively freezes every v8 object on |object|.
+bool DeepFreeze(const v8::Local<v8::Object>& object,
+                const v8::Local<v8::Context>& context,
+                std::set<int> frozen = std::set<int>()) {
+  int hash = object->GetIdentityHash();
+  if (frozen.find(hash) != frozen.end())
+    return true;
+  frozen.insert(hash);
+
+  v8::Local<v8::Array> property_names =
+      object->GetOwnPropertyNames(context).ToLocalChecked();
+  for (uint32_t i = 0; i < property_names->Length(); ++i) {
+    v8::Local<v8::Value> child =
+        object->Get(context, property_names->Get(context, i).ToLocalChecked())
+            .ToLocalChecked();
+    if (child->IsObject() && !child->IsTypedArray()) {
+      if (!DeepFreeze(v8::Local<v8::Object>::Cast(child), context, frozen))
+        return false;
+    }
+  }
+  return mate::internal::IsTrue(
+      object->SetIntegrityLevel(context, v8::IntegrityLevel::kFrozen));
+}
+
+bool IsPlainObject(const v8::Local<v8::Value>& object) {
+  if (!object->IsObject())
+    return false;
+
+  return !(object->IsNullOrUndefined() || object->IsDate() ||
```

Comment:
Can you try upstreaming this utility to `v8.h` , manually tracking this variation everytime is not good. Maybe upstream can suggest other alternatives to test this.

Reply:
Yeah can do 👍